File: src/main/java/skytils/skytilsmod/features/impl/events/TechnoMayor.java
package skytils.skytilsmod.features.impl.events;

import com.google.common.collect.ImmutableList;
import net.minecraft.client.Minecraft;
import net.minecraft.client.renderer.GlStateManager;
import net.minecraft.entity.Entity;
import net.minecraft.entity.EntityLivingBase;
import net.minecraft.entity.item.EntityArmorStand;
import net.minecraft.network.play.server.S02PacketChat;
import net.minecraft.util.AxisAlignedBB;
import net.minecraft.util.BlockPos;
import net.minecraft.util.Vec3;
import net.minecraftforge.client.event.RenderLivingEvent;
import net.minecraftforge.client.event.RenderWorldLastEvent;
import net.minecraftforge.event.world.WorldEvent;
import net.minecraftforge.fml.common.eventhandler.EventPriority;
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent;
import skytils.skytilsmod.Skytils;
import skytils.skytilsmod.events.PacketEvent;
import skytils.skytilsmod.utils.*;

import java.awt.*;
import java.util.ArrayList;
import java.util.List;

import static skytils.skytilsmod.Skytils.mc;

public class TechnoMayor {

    private static final List<Vec3> orbLocations = new ArrayList<>();

    @SubscribeEvent
    public void onRenderSpecialLivingPre(RenderLivingEvent.Specials.Pre<EntityLivingBase> event) {
        if(!Utils.inSkyblock) return;
        Entity entity = event.entity;

        if (!(event.entity instanceof EntityArmorStand) || !entity.hasCustomName() ||event.entity.isDead) return;
        EntityArmorStand e = (EntityArmorStand) entity;
        if (!e.getCustomNameTag().equals("§6§lSHINY ORB")) return;

        BlockPos origin = e.getPosition();
        List<Entity> nearbyEntities = mc.theWorld.getEntitiesWithinAABBExcludingEntity(e, new AxisAlignedBB(origin, new BlockPos(origin.getX() + 1, origin.getY() + 1, origin.getZ() + 1)));
        for(Entity ent : nearbyEntities) {
            if(ent instanceof EntityArmorStand && ent.hasCustomName() && ent.getCustomNameTag().contains(mc.thePlayer.getName())) {
                ent.worldObj.removeEntity(ent);
                e.worldObj.removeEntity(e);
                orbLocations.add(new Vec3(origin.getX() + 0.5, origin.getY() - 2, origin.getZ() + 0.5));
                break;
            }
        }

    }

    @SubscribeEvent
    public void onWorldRender(RenderWorldLastEvent event) {
        if (!Utils.inSkyblock) return;
        if (SBInfo.getInstance().getLocation() == null || !SBInfo.getInstance().getLocation().equalsIgnoreCase("hub")) return;
        if (!Skytils.config.shinyOrbWaypoints) return;
        Entity viewer = Minecraft.getMinecraft().getRenderViewEntity();
        double viewerX = viewer.lastTickPosX + (viewer.posX - viewer.lastTickPosX) * event.partialTicks;
        double viewerY = viewer.lastTickPosY + (viewer.posY - viewer.lastTickPosY) * event.partialTicks;
        double viewerZ = viewer.lastTickPosZ + (viewer.posZ - viewer.lastTickPosZ) * event.partialTicks;

        for (Vec3 orb : ImmutableList.copyOf(orbLocations)) {
            double x = orb.xCoord - viewerX;
            double y = orb.yCoord - viewerY;
            double z = orb.zCoord - viewerZ;
            double distSq = x*x + y*y + z*z;

            GlStateManager.disableCull();
            GlStateManager.disableTexture2D();
            if (distSq > 5*5) RenderUtil.renderBeaconBeam(x, y, z, new Color(114, 245, 82).getRGB(), 0.75f, event.partialTicks);
            GlStateManager.disableDepth();
            RenderUtil.renderWaypointText("Orb", orb.xCoord, orb.yCoord + 1.5f, orb.zCoord, event.partialTicks);
            GlStateManager.disableLighting();
            GlStateManager.enableTexture2D();
            GlStateManager.enableDepth();
            GlStateManager.enableCull();
        }
    }

    @SubscribeEvent(priority = EventPriority.LOWEST, receiveCanceled = true)
    public void onChatPacket(PacketEvent.ReceiveEvent event) {
        if (!(event.packet instanceof S02PacketChat)) return;
        S02PacketChat packet = (S02PacketChat) event.packet;
        if (packet.getType() == 2) return;
        String unformatted = StringUtils.stripControlCodes(packet.getChatComponent().getUnformattedText());
        String formatted = packet.getChatComponent().getFormattedText();

        if(unformatted.equals("Your Shiny Orb and associated pig expired and disappeared.") || unformatted.equals("SHINY! The orb is charged! Click on it for loot!")) {
            orbLocations.removeIf(pos -> mc.thePlayer.getPosition().distanceSq(pos.xCoord, pos.yCoord, pos.zCoord) < 7 * 7);
        }
    }

    @SubscribeEvent
    public void onWorldChange (WorldEvent.Load event) {
        orbLocations.removeIf(e -> true);
    }
}

File: src/main/kotlin/skytils/skytilsmod/features/impl/events/TechnoMayor.kt
/*
 * Skytils - Hypixel Skyblock Quality of Life Mod
 * Copyright (C) 2021 Skytils
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published
 * by the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
package skytils.skytilsmod.features.impl.events

import com.google.common.collect.ImmutableList
import net.minecraft.client.Minecraft
import net.minecraft.client.renderer.GlStateManager
import net.minecraft.entity.Entity
import net.minecraft.entity.EntityLivingBase
import net.minecraft.entity.item.EntityArmorStand
import net.minecraft.network.play.server.S02PacketChat
import net.minecraft.util.AxisAlignedBB
import net.minecraft.util.BlockPos
import net.minecraft.util.Vec3
import net.minecraftforge.client.event.RenderLivingEvent
import net.minecraftforge.client.event.RenderWorldLastEvent
import net.minecraftforge.event.world.WorldEvent
import net.minecraftforge.fml.common.eventhandler.EventPriority
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent
import skytils.skytilsmod.Skytils
import skytils.skytilsmod.events.PacketEvent.ReceiveEvent
import skytils.skytilsmod.utils.*
import java.awt.Color

class TechnoMayor {
    @SubscribeEvent
    fun onRenderSpecialLivingPre(event: RenderLivingEvent.Specials.Pre<EntityLivingBase?>) {
        if (!Utils.inSkyblock) return
        val entity: Entity = event.entity
        if (event.entity !is EntityArmorStand || !entity.hasCustomName() || event.entity.isDead) return
        val e = entity as EntityArmorStand
        if (e.customNameTag != "§6§lSHINY ORB") return
        val origin = e.position
        val nearbyEntities = mc.theWorld.getEntitiesWithinAABBExcludingEntity(
            e,
            AxisAlignedBB(origin, BlockPos(origin.x + 1, origin.y + 1, origin.z + 1))
        )
        for (ent in nearbyEntities) {
            if (ent is EntityArmorStand && ent.hasCustomName() && ent.getCustomNameTag().contains(mc.thePlayer.name)) {
                ent.worldObj.removeEntity(ent)
                e.worldObj.removeEntity(e)
                orbLocations.add(Vec3(origin.x + 0.5, (origin.y - 2).toDouble(), origin.z + 0.5))
                break
            }
        }
    }

    @SubscribeEvent
    fun onWorldRender(event: RenderWorldLastEvent) {
        if (!Utils.inSkyblock) return
        if (SBInfo.mode != SBInfo.SkyblockIslands.HUB.mode && SBInfo.mode != SBInfo.SkyblockIslands.FARMINGISLANDS.mode) return
        if (!Skytils.config.shinyOrbWaypoints) return

        val (viewerX, viewerY, viewerZ) = RenderUtil.getViewerPos(event.partialTicks)
        for (orb in ImmutableList.copyOf(orbLocations)) {
            val x = orb.xCoord - viewerX
            val y = orb.yCoord - viewerY
            val z = orb.zCoord - viewerZ
            val distSq = x * x + y * y + z * z
            GlStateManager.disableCull()
            GlStateManager.disableTexture2D()
            if (distSq > 5 * 5) RenderUtil.renderBeaconBeam(x, y, z, Color(114, 245, 82).rgb, 0.75f, event.partialTicks)
            GlStateManager.disableDepth()
            RenderUtil.renderWaypointText("Orb", orb.xCoord, orb.yCoord + 1.5f, orb.zCoord, event.partialTicks)
            GlStateManager.disableLighting()
            GlStateManager.enableTexture2D()
            GlStateManager.enableDepth()
            GlStateManager.enableCull()
        }
    }

    @SubscribeEvent(priority = EventPriority.LOWEST, receiveCanceled = true)
    fun onChatPacket(event: ReceiveEvent) {
        if (event.packet !is S02PacketChat) return
        val packet = event.packet
        if (packet.type.toInt() == 2) return
        val unformatted = StringUtils.stripControlCodes(packet.chatComponent.unformattedText)
        if (unformatted == "Your Shiny Orb and associated pig expired and disappeared." || unformatted == "SHINY! The orb is charged! Click on it for loot!") {
            orbLocations.removeIf { pos: Vec3 ->
                mc.thePlayer.position.distanceSq(
                    pos.xCoord,
                    pos.yCoord,
                    pos.zCoord
                ) < 7 * 7
            }
        }
    }

    @SubscribeEvent
    fun onWorldChange(event: WorldEvent.Load?) {
        orbLocations.clear()
    }

    companion object {
        private val mc = Minecraft.getMinecraft()
        private val orbLocations: MutableList<Vec3> = ArrayList()
    }
}